pkg/s3wrapper/filesystem_test.go

```diff
@@ -194,18 +202,162 @@ var _ = Describe("s3filesystem", func() {
+				"key3": "bar",
+				"key5": "foo",
+			})).To(BeFalse())
+		}
+
+		It("ListObjectsByPrefix returns metadata about objects", func() {
+			createFileObject(client.basedir, fmt.Sprintf("dir/other/%s", uuid.New().String()), now, map[string]string{
+				"Key1": "bar",
+				"key2": "foo",
+			})
+			createFileObject(client.basedir, fmt.Sprintf("dir/other/%s", uuid.New().String()), now, map[string]string{
+				"key3": "bar2",
+				"Key4": "foo2",
+			})
+			validateListObjectsByPrefix()
+		})
+
+		Context("Upload", func() {
+
+			var (
+				file1Path         string
+				file2Path         string
+				tempFileDirectory string
+			)
+
+			BeforeEach(func() {
+				var err error
+				tempFileDirectory, err = os.MkdirTemp(os.TempDir(), fmt.Sprintf("manifest-test-%s", uuid.NewString()))
+				Expect(err).ToNot(HaveOccurred())
+				file1Path, _ = createFileObject(tempFileDirectory, uuid.NewString(), now, nil)
+				file2Path, _ = createFileObject(tempFileDirectory, uuid.NewString(), now, nil)
+			})
+
+			AfterEach(func() {
+				Expect(os.Remove(file1Path)).To(BeNil())
+				Expect(os.Remove(file2Path)).To(BeNil())
+				Expect(os.Remove(tempFileDirectory)).To(BeNil())
+			})
+
+			It("Upload stores metadata about objects", func() {
+				dataStr := "hello world"
+				err := client.UploadWithMetadata(ctx, []byte(dataStr), "dir/other/file1", map[string]string{
+					"Key1": "bar",
+					"key2": "foo",
+				})
+				Expect(err).Should(BeNil())
+				err = client.UploadWithMetadata(ctx, []byte(dataStr), "dir/other/file2", map[string]string{
+					"key3": "bar2",
+					"Key4": "foo2",
+				})
+				Expect(err).Should(BeNil())
+				validateListObjectsByPrefix()
+			})
+
+			It("UploadFile stores metadata about objects", func() {
+				Expect(client.UploadFileWithMetadata(ctx, file1Path, "dir/other/file1", map[string]string{
+					"Key1": "bar",
+					"key2": "foo",
+				})).To(BeNil())
+				Expect(client.UploadFileWithMetadata(ctx, file2Path, "dir/other/file2", map[string]string{
+					"key3": "bar2",
+					"Key4": "foo2",
+				})).To(BeNil())
+				validateListObjectsByPrefix()
+			})
+
+			It("UploadStream stores metadata about objects", func() {
+				fileReader, err := os.Open(file1Path)
+				Expect(err).Should(BeNil())
+				err = client.UploadStreamWithMetadata(ctx, fileReader, "dir/other/file1", map[string]string{
+					"Key1": "bar",
+					"key2": "foo",
+				})
+				Expect(err).Should(BeNil())
+				fileReader.Close()
+
+				fileReader, err = os.Open(file2Path)
+				Expect(err).Should(BeNil())
+				err = client.UploadStreamWithMetadata(ctx, fileReader, "dir/other/file2", map[string]string{
+					"key3": "bar2",
+					"Key4": "foo2",
+				})
+				Expect(err).Should(BeNil())
+				fileReader.Close()
+
+				validateListObjectsByPrefix()
+			})
+		})
 	})
 
 	AfterEach(func() {
 		os.RemoveAll(baseDir)
 	})
 })
 
-func createFileObject(baseDir, objKey string, imgCreatedAt time.Time) (string, os.FileInfo) {
+func ignoringEINTR(fn func() error) (err error) {
```

Comment:
Why do we need to do this?

Reply:
I looked at the way the xattr library implements the writing of xattr attributes and one of the problems that can be encountered during writing is an "interrupted exception"

https://github.com/pkg/xattr/blob/master/xattr_linux.go#L26C1-L39

Though reading closer, this only applies to FUSE and CIFS, so I can probably remove this.